src/app/zap_cluster_list.py

```diff
@@ -0,0 +1,146 @@
+#!/usr/bin/env python3
+"""Parses a ZAP input file and outputs directories to compile."""
+
+import argparse
+import pathlib
+import json
+import typing
+import sys
+
+# List of directories in src/app/clusters to build for server clusters.
+SERVER_CLUSTERS: typing.Dict[str, typing.List[str]] = {
+    'ALARM_CLUSTER': [],
+    'APPLICATION_BASIC_CLUSTER': [],
+    'BARRIER_CONTROL_CLUSTER': ['barrier-control-server'],
+    'BASIC_CLUSTER': ['basic'],
+    'BINDING_CLUSTER': ['bindings'],
+    'COLOR_CONTROL_CLUSTER': ['color-control-server'],
+    'COMMISSIONING_CLUSTER': [],
+    'CONTENT_LAUNCH_CLUSTER': ['content-launch-server'],
+    'DESCRIPTOR_CLUSTER': ['descriptor'],
+    'DEVICE_TEMP_CLUSTER': [],
+    'DOOR_LOCK_CLUSTER': ['door-lock-server'],
+    'GENERAL_COMMISSIONING_CLUSTER': ['general-commissioning-server'],
+    'GROUPS_CLUSTER': ['groups-server'],
+    'GROUP_KEY_MANAGEMENT_CLUSTER': [],
+    'IAS_ZONE_CLUSTER': ['ias-zone-server'],
+    'IDENTIFY_CLUSTER': ['identify'],
+    'LEVEL_CONTROL_CLUSTER': ['level-control'],
+    'LOW_POWER_CLUSTER': ['low-power-server'],
+    'MEDIA_PLAYBACK_CLUSTER': ['media-playback-server'],
+    'NETWORK_COMMISSIONING_CLUSTER': ['network-commissioning'],
+    'OCCUPANCY_SENSING_CLUSTER': [],
+    'ON_OFF_CLUSTER': ['on-off-server'],
+    'OTA_BOOTLOAD_CLUSTER': [],
+    'POWER_CONFIG_CLUSTER': [],
+    'SCENES_CLUSTER': ['scenes'],
+    'TEMP_MEASUREMENT_CLUSTER': ['temperature-measurement-server'],
+    'WINDOW_COVERING_CLUSTER': [],
+    'ZLL_COMMISSIONING_CLUSTER': []
+}
+
+# List of directories in src/app/clusters to build for client clusters.
+CLIENT_CLUSTERS: typing.Dict[str, typing.List[str]] = {
+    'ALARM_CLUSTER': [],
+    'APPLICATION_BASIC_CLUSTER': [],
+    'BARRIER_CONTROL_CLUSTER': [],
+    'BASIC_CLUSTER': [],
+    'BINDING_CLUSTER': [],
+    'COLOR_CONTROL_CLUSTER': [],
+    'COMMISSIONING_CLUSTER': [],
+    'CONTENT_LAUNCH_CLUSTER': [],
+    'DEVICE_TEMP_CLUSTER': [],
+    'DESCRIPTOR_CLUSTER': [],
+    'DOOR_LOCK_CLUSTER': ['door-lock-client'],
+    'GENERAL_COMMISSIONING_CLUSTER': [],
+    'GROUPS_CLUSTER': ['groups-client'],
+    'GROUP_KEY_MANAGEMENT_CLUSTER': [],
+    'IAS_ZONE_CLUSTER': ['ias-zone-client'],
+    'IDENTIFY_CLUSTER': [],
+    'LEVEL_CONTROL_CLUSTER': [],
+    'LOW_POWER_CLUSTER': [],
+    'MEDIA_PLAYBACK_CLUSTER': ['media-playback-client'],
+    'NETWORK_COMMISSIONING_CLUSTER': [],
+    'OCCUPANCY_SENSING_CLUSTER': [],
+    'ON_OFF_CLUSTER': [],
+    'OTA_BOOTLOAD_CLUSTER': [],
+    'POWER_CONFIG_CLUSTER': [],
+    'SCENES_CLUSTER': ['scenes-client'],
+    'TEMP_MEASUREMENT_CLUSTER': [],
+    'WINDOW_COVERING_CLUSTER': [],
+    'ZLL_COMMISSIONING_CLUSTER': []
+}
```

Comment:
I guess this is fine at the moment, but it seems painful to have to maintain those lists.

For the future, the folder names should be normalised  so the script will just check if a directory exists.
And the script may fail if a directory is required but does not exists. For example, a cluster directory is required if the cluster has any enable non-manufacturer specific command (because the code for the command callback should be implemented somewhere).

Reply:
Checking if a directory exists won't work correctly for incremental builds.

Reply:
ZAP could presumably generate these maps, though.